src/core/lib/iomgr/vsock.h

```diff
@@ -0,0 +1,40 @@
+//
+//
+// Copyright 2021 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#ifndef GRPC_SRC_CORE_LIB_IOMGR_VSOCK_H
+#define GRPC_SRC_CORE_LIB_IOMGR_VSOCK_H
+
+#include <grpc/support/port_platform.h>
+
+#include <string>
+
+#include "absl/strings/string_view.h"
+
+#include <grpc/support/string_util.h>
+
+#include "src/core/lib/iomgr/port.h"
+#include "src/core/lib/iomgr/resolve_address.h"
+
+void grpc_create_socketpair_if_vsock(int sv[2]);
```

Comment:
@ctiller Do you think it makes sense to have a new core e2e test fixture for a vsock socket pair?

I think the existing core e2e test fixture for UDS socket pairs is the only reason that we have the equivalent of this function for UDS sockets.  If we're not going to add such a fixture, then we probably don't need this function.

Reply:
> Overall, this looks good! Comments are mostly minor.
> 
> Looks like there's a merge conflict that needs to be resolved. Please merge master into this branch rather than force-pushing. Thanks!
> 
> @drfloob, can you please provide a second review for this?

Thank you @markdroth. I have resolved most comments. For the e2e test fixture, I think I need comments from @ctiller ?

Reply:
Sorry, I missed your request to merge instead of force-push, I used to use rebase and force-push...... I will follow the rule for further commits.

Reply:
If we don't add an e2e test for it, I'm not sure how we'd keep this code path tested.

Reply:
Yeah I think we'll want an e2e test fixture for this - to make sure all the bits work together (I'd love separate unit tests for the resolver, socket management, proxy hooks, etc.. but we don't have that yet for anything else)

Reply:
What is the test infra like for this? vsock requires specific kernel configs to be enabled, so a VM with a properly configured kernel might be needed (unless the infra has the right configs available).

Reply:
There is an added hiccup here: gRPC supports CentOS 7, which distributes linux kernel v3.10. I don't think this PR will build on CentOS 7 right now. `GRPC_HAVE_VSOCK` will have to be defined more carefully: only if we can detect that the current platform supports it. See https://opensource.google/documentation/policies/cplusplus-support

The path to testing this may be a bit messy. It would require an entirely new job to be set up similar to the test configs in `tools/internal_ci/linux/*`, with some corresponding internal Google changes to make the test run (that last bit is not hard). It's possible for you to do, but it would be some extra work to grok the system. Also, the current `socketpair` implementation in this PR won't work, per https://github.com/torvalds/linux/blob/47a2ee5d4a0bda05decdda7be0a77e792cdb09a3/net/vmw_vsock/af_vsock.c#L2225, so something like VSOCK-local would have to be used to mimic a socket pair, so the end2end tests can run. See here for an example of the end2end test fixture for other socketpairs https://github.com/grpc/grpc/blob/0d678a9551608215c95f755b3562e90589a4fc3e/test/core/end2end/end2end_test_suites.cc#L729-L755

If the fixes can be made to get this to build cleanly on CentOS 7, I'd be in support of documenting this everywhere as community-supported and leave it mostly untested. That does mean that if it breaks (and we won't know it when it happens), the community would need to repair it, or the functionality would likely be removed.

Reply:
@drfloob, 
1. for CentOS 7(CentOS Linux release 7.9.2009 (Core)), I have installed the OS and tried to build gRPC on it, the result turns out: 

- **Failed** to build with bazel(w or w/o this PR), see below errors:
```
[sdp@a4bf016a433b grpc]$ bazel build :all
INFO: Running bazel wrapper (see //tools/bazel for details), bazel version 6.1.2 will be used instead of system-wide bazel installation.
Starting local Bazel server and connecting to it...
INFO: Analyzed 174 targets (169 packages loaded, 5375 targets configured).
INFO: Found 174 targets...
INFO: From Compiling utf8_validity.cc [for tool]:
external/utf8_range/utf8_validity.cc:139:12: warning: unused function 'CodepointSkipBackwards' [-Wunused-function]
inline int CodepointSkipBackwards(int32_t codepoint_word) {
           ^
1 warning generated.
ERROR: /home/sdp/grpc/BUILD:677:16: Compiling src/core/lib/gpr/alloc.cc failed: (Exit 1): clang-11 failed: error executing command (from target //:gpr) /opt/rh/llvm-toolset-11.0/root/usr/bin/clang-11 -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer '-std=c++0x' -MD -MF ... (remaining 75 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
src/core/lib/gpr/alloc.cc:21:10: error: module //:gpr does not depend on a module exporting 'stdlib.h'
#include <stdlib.h>
         ^
1 error generated.
INFO: Elapsed time: 27.790s, Critical Path: 5.43s
INFO: 928 processes: 765 internal, 163 processwrapper-sandbox.
FAILED: Build did NOT complete successfully
```
- **Succeed** to build with cmake(w or w/o this PR).
- Since the build issue with bazel is not introduced by this PR, we can ignore the failure in this session, right?
2. Just as you commented, we need to define `GRPC_HAVE_VSOCK` carefully, I have uploaded new commits to guard `GRPC_HAVE_VSOCK`, it will be enabled only when build with linux version >= 3.9. And also removed `grpc_create_socketpair_if_vsock()` since it is not supported. please help review.

Reply:
> Failed to build with bazel(w or w/o this PR), see below errors:

Oh, fun! That looks like a bazel configuration issue. What version of bazel did you use? gRPC only supports 5.x and newer https://github.com/grpc/grpc/blob/master/bazel/supported_versions.txt. Something like [bazelisk](https://github.com/bazelbuild/bazelisk) would let you test with a supported version, if that's the issue.

> I have uploaded new commits to guard GRPC_HAVE_VSOCK, it will be enabled only when build with linux version >= 3.9.

That looks good. I've kicked off the CI again, and I'll take another pass at your implementation today. Thank you for your persistence!

Reply:
> What version of bazel did you use? gRPC only supports 5.x and newer https://github.com/grpc/grpc/blob/master/bazel/supported_versions.txt. Something like [bazelisk](https://github.com/bazelbuild/bazelisk) would let you test with a supported version, if that's the issue.

Original bazel version is v4.2.1, but even after I upgrade it to v6.1.2, the issue still exists.
```
[sdp@a4bf016a433b grpc]$ bazel --version
bazel 6.1.2
[sdp@a4bf016a433b grpc]$ bazel build :all
Starting local Bazel server and connecting to it...
INFO: Analyzed 174 targets (169 packages loaded, 5375 targets configured).
INFO: Found 174 targets...
ERROR: /home/sdp/grpc/BUILD:2408:16: Compiling src/core/lib/config/config_vars_non_generated.cc failed: (Exit 1): clang-11 failed: error executing command (from target //:config_vars) /opt/rh/llvm-toolset-11.0/root/usr/bin/clang-11 -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer '-std=c++0x' -MD -MF ... (remaining 57 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
src/core/lib/config/config_vars_non_generated.cc:19:10: error: module //:config_vars does not depend on a module exporting 'atomic'
#include <atomic>
         ^
src/core/lib/config/config_vars_non_generated.cc:20:10: error: module //:config_vars does not depend on a module exporting 'string'
#include <string>
         ^
2 errors generated.
INFO: Elapsed time: 25.616s, Critical Path: 3.96s
INFO: 1016 processes: 884 internal, 132 processwrapper-sandbox.
FAILED: Build did NOT complete successfully
```
I think we can create a seperate issue for the grpc build with bazel on CentOS 7, no need to expand more here.

For the CI results, there still 3 failures:
[Bazel BinderTransport APK building/testing](https://source.cloud.google.com/results/invocations/b6d4559e-6b43-49db-bb9e-778a4a998348) : submitted commit to resolve it
[Distribution Tests Python Linux](https://source.cloud.google.com/results/invocations/88f6f3fc-c2ab-4eb5-9b01-3745ce4c9150) : No idea...
[IWYU (internal CI) ](https://source.cloud.google.com/results/invocations/d8cab4bc-f25a-4990-9237-373b5220c653) : It seems suggest some including position adjust, not sure to follow, do you have any comments?

Reply:
> I think we can create a seperate issue for the grpc build with bazel on CentOS 7, no need to expand more here.

I totally agree, let's remove that requirement from this work, we'll have to work on it separately. Thank you for checking.

Regarding the other issues, I've sent a PR to fix the IWYU problems, the tool needed a little guidance: https://github.com/YadongQi/grpc/pull/22. The python problem is unrelated, and won't block this PR.